rust/kernel/platdev.rs

```diff
@@ -18,6 +21,7 @@ use core::{marker::PhantomPinned, pin::Pin};
 #[derive(Default)]
 pub struct Registration {
     registered: bool,
+    of_table: Option<*const c_types::c_void>,
```

Comment:
I think this can be an `Option<NonNull<c_types::c_void>>` because as far as I can see all pointers it could contain ultimately come from a `Box` allocation. (This would safe at least one byte for the `Option` discriminant)

Reply:
As far as I can see, creating a `NonNull` from a `ptr` requires one additional `unwrap` or `Err()` return, which would introduce one more `// PANIC:` annotation?
```rust
pub fn new(ptr: *mut T) -> Option<NonNull<T>>
```
Also its docs say `but is ultimately more dangerous to use because of its additional properties` ?
So not sure if the pros outweigh the cons here.
But thank you for introducing me to `NonNull` ! It's basically a covariant pointer type, I didn't even know that this existed. I'm sure it'll be useful in many other places.

Reply:
I suspect that in this case (a one-off use inside `Registration`), the light is probably not worth the candle. But, `NonNull<T>`  might be extremely useful as part of the implementation detail of more general, core abstractions. The more such an abstraction is re-used, the more we can accept verbosity and "tricky" guarantees inside of it.

I really appreciate your knowledge and expertise !

Reply:
I had assumed that the `Registration` would live for the whole lifetime of the module so saving up to 8 bytes for basically free would be worth it but I agree that in this case it probably doesn't matter that much.

> I really appreciate your knowledge and expertise!

I'm glad I can at least contribute something on the Rust usage side since I have very little prior experience with linux kernel code.

Reply:
> But, `NonNull<T>` might be extremely useful as part of the implementation detail of more general, core abstractions.

Would it maybe make sense to encourage using this by providing a helper method like this:
```rust
/// Attempts to allocate `x` on the heap using a `Box` and provides a pointer to that allocation.
fn bikeshed_alloc<T>(x: T) -> Result<NonNull<T>> {
    NonNull::from(Box::leak(Box::try_new(x)?))
}
```

This would remove most of the complexity around using `NonNull`, deduplicate a lot of the `Box::into_raw(Box::try_new(...)?)` logic and would of course still allow people to use `.as_ptr()` to get a "normal" raw pointer if they need to e.g. do arithmetic with it.

Reply:
Ah I had not yet encountered `PointerWrapper`. I agree that it doesn't really make sense to provide a helper orthogonal to that.

Reply:
`NonNull<T>` may make sense in some of the code surrounding `PointerWrapper - when there's a case to refine that abstraction.

Reply:
`NonNull` is generally preferred in data structures wherever a pointer is sure to be not null. `Option<NonNull<T>>` can be used if a pointer is nullable. Note that `Option<NonNull<T>>` is even FFI-safe -- it is treated exactly as `*mut T` for ABI purporses.

Doing so forces the user to consider about the nullability of a pointer, so it can prevent null dereference even in unsafe code. The only exception are probably local variables, as raw pointers have more helper function to do arithmetic on.